src/libsystemd/sd-daemon/sd-daemon.c

```diff
@@ -467,13 +467,34 @@ _public_ int sd_pid_notify_with_fds(
         if (!e)
                 return 0;
 
-        r = sockaddr_un_set_path(&sockaddr.un, e);
+        /* Allow AF_UNIX and AF_VSOCK, reject the rest. */
+        r = socket_local_address_parse(&address, e);
         if (r < 0)
                 goto finish;
-        msghdr.msg_namelen = r;
+        msghdr.msg_namelen = address.size;
+
+        /* If we didn't get an address, talk to the hypervisor by default, as that's the purpose of this functionality. */
+        if (address.sockaddr.vm.svm_family == AF_VSOCK && address.sockaddr.vm.svm_cid == VMADDR_CID_ANY)
+                address.sockaddr.vm.svm_cid = VMADDR_CID_HYPERVISOR;
+
+        /* At the time of writing QEMU does not yet support AF_VSOCK + SOCK_DGRAM and returns
+         * ENODEV. Fallback to SOCK_SEQPACKET in that case. */
+        fd = socket(address.sockaddr.sa.sa_family, SOCK_DGRAM|SOCK_CLOEXEC, 0);
+        if (fd < 0 && errno == ENODEV && address.sockaddr.sa.sa_family == AF_VSOCK) {
+                fd = socket(address.sockaddr.sa.sa_family, SOCK_SEQPACKET|SOCK_CLOEXEC, 0);
+                if (fd < 0) {
+                        r = -errno;
+                        goto finish;
+                }
+
+                if (connect(fd, &address.sockaddr.sa, address.size) < 0) {
+                        r = -errno;
+                        goto finish;
+                }
```

Comment:
hmm, given that the receiver on the host and the hypervisor are two different things and might support different functionality, shouldn't we try SOCK_SEQPACKET also if we can allocate SOCK_DGRAM but not connect() to it? i.e. let's say one day qemu learns SOCK_DGRAM, but libvirt still only does SOCK_SEQPACKET, then you'd not be able to communicate with libvirt's receiver anymore...

Reply:
I'm not sure I follow here, there's no connect in the dgram case? The problem is the vsock driver implementation in the hypervisor, that's what's missing support for DGRAM, and looking at the attempts to add it, it doesn't look like a trivial feature to add.
On the other hand, for a consumer simply using AF_VSOCK it's trivial to switch from one protocol to the other.

I figure a consumer should know what the hypervisor is going to support (eg: vmware supports DGRAM as far as I can read), so it can make the right choice.

Reply:
again, in a libvirt/qemu world I'd assume that libvirt listens on the AF_VSOCK thing, and qemu does not. But that means qemu and libvirt must agree on the same socket type, because otherwise qemu might open something up that libvirt isn't listening on.

i mean, you are right, forget my comment about connect(), it's really sendmsg() i guess... i.e. if sendmsg() fails on AF_VSOCK/SOCK_DGRAM even though we managed to allocate a socket for that, maybe retry immediately with AF_VSOCK/SOCK_SEQPACKET? because in that case libvirt (as the listener) and qemu (as the transport) can enable support for SOCK_DGRAM independently of each other and things will still work correctly. because otherwise, if qemu starts supporting SOCK_DGRAM but libvirt hasn't been updated yet we'll send messages onto a socket that can't access it.

Reply:
i am not sure what "is trivial" has to do with anything. I am just saying that qemu/libvirt might not be updated in lockstep, and hence a robust fallback would make sense.

libvirt implementing both mechanisms right now is problematic, since they can't test it, since they only support qemu and qemu doesn't support SOCK_DGRAM.

This is actually similar for us: we are adding support for SOCK_DGRAM without actually being able to test this. But I think we have a stronger incentive to risk it, since we want compat with vmware and everything else that supports SOCK_DGRAM.

Hence, I am pretty sure that qemu will start supporting SOCK_DGRAM before libvirt will start making use of it, hence I think we should think about the compat strategy here...

Reply:
Mh? libvirt supports a gazillion hypervisors, not just qemu, including vmware (and xen, and virtualbox, and hyperv, etc): https://en.wikipedia.org/wiki/Libvirt#Supported_Hypervisors
Hence why if it ever consumes this, it has to do it with both protocols from day one, there's really no other option due to the dependency on the hypervisor + guest kernel.